src/Layers/ThrottleLayer.lua

```diff
@@ -25,24 +27,40 @@ function ThrottleLayer._perform(methodName, collectionName, ...)
 	if ThrottleLayer._queue[resource] == nil then
 		ThrottleLayer._queue[resource] = {}
 
-		local connection
-		connection = RunService.Heartbeat:Connect(function()
-			for _, thread in ipairs(ThrottleLayer._queue[resource]) do
-				if DataStoreService:GetRequestBudgetForRequestType(resource) > 0 then
-					DataStoreLayer.perform(unpack(thread))
+		Promise.defer(function(resolve)
+			while #ThrottleLayer._queue[resource] > 0 do
+				local request = table.remove(ThrottleLayer._queue[resource], 1)
+
+				while DataStoreService:GetRequestBudgetForRequestType(resource) == 0 do
+					RunService.Heartbeat:Wait()
+				end
+
+				local ok, result = pcall(DataStoreLayer.perform, unpack(request.args))
+				if ok then
+					request.resolve(result)
 				else
-					break
+					request.reject(Error.new({
+						kind = Error.Kind.DataStoreError,
+						error = result
+					}))
 				end
 			end
 
-			if #ThrottleLayer._queue[resource] == 0 then
-				connection:Disconnect()
-				ThrottleLayer._queue[resource] = nil
-			end
+			ThrottleLayer._queue[resource] = nil
+			resolve()
```

Comment:
Can we switch this `Promise.defer` out for a:

```lua
coroutine.wrap(function()
  RunService.Hearbeat:Wait()
  ...other code here...
end)()
```

This feels cleaner to me, as we aren't using any of the features of a Promise here.